libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,206 @@
+	m_currentNode = SpecialNode::EntryCreation;
+	m_currentDispatch = SpecialNode::InternalCreationDispatch;
+
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts | boost::adaptors::reversed)
+	{
+		for (auto const* stateVar: contract->stateVariables())
+			stateVar->accept(*this);
+
+		for (auto arg: contract->baseContracts())
+			arg->accept(*this);
+
+		if (contract->constructor())
+		{
+			add(*m_currentNode, contract->constructor());
+			contract->constructor()->accept(*this);
+			m_currentNode = contract->constructor();
+		}
+	}
+
+	m_currentNode.reset();
+	m_currentDispatch = SpecialNode::InternalDispatch;
+
+	// Create graph for all publicly reachable functions
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			if (!m_graph->edges.count(funcDef))
+				visitCallable(funcDef);
+
+		// Add all external functions to the RuntimeDispatch
+		add(SpecialNode::Entry, &functionType->declaration());
+	}
+
+	// Add all InternalCreationDispatch calls to the RuntimeDispatch as well
+	add(SpecialNode::InternalDispatch, SpecialNode::InternalCreationDispatch);
+
+	if (_contract.fallbackFunction())
+		add(SpecialNode::Entry, _contract.fallbackFunction());
+
+	if (_contract.receiveFunction())
+		add(SpecialNode::Entry, _contract.receiveFunction());
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, Node const& _rhs) const
+{
+	if (_lhs.index() != _rhs.index())
+		return _lhs.index() < _rhs.index();
+
+	if (std::holds_alternative<SpecialNode>(_lhs))
+		return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+	return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, int64_t _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+	return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(int64_t _lhs, Node const& _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+	return _lhs < std::get<ASTNode const*>(_rhs)->id();
+}
+
+unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	return FunctionCallGraphBuilder(_contract).m_graph;
+}
+
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		solAssert(*_identifier.annotation().requiredLookup == VirtualLookup::Virtual, "");
+
+		auto funType = dynamic_cast<FunctionType const*>(_identifier.annotation().type);
+
+		// For events kind() == Event, so we have an extra check here
+		if (funType && funType->kind() == FunctionType::Kind::Internal)
```

Comment:
Can this be `Kind::Declaration` here? If not, can you add an assert?

Reply:
I think it can't. At least none of my tests triggers that line when I change it to `Kind::Declaration`.

Actually what does `Kind::Declaration` represent?